Add syntax highlighting for vim and neovim

x2w-soda opened the pull request:
PR to add a vim syntax file to support syntax highlighting for .carbon files. Since the Carbon language specification is still in progress, a single syntax file is easier to maintain and adapt to design changes in Carbon. After the language syntax has matured and stabilized, advanced tooling such as Treesitter and Language Servers can take over the syntax highlighting in neovim.

### Highlighting Support
- comments and preprocessors (RUN, CHECK, etc)
- string and numeric literals
- primitive type names, classes, aliases
- control flow constructs
- identifier names
- package and library declarations
- most keywords mentioned in `docs/design/README.md`

### With a Dark Colorscheme:
![carbon01](https://user-images.githubusercontent.com/78875280/181017032-e2c140c8-e98a-46d3-8f94-b8dc0f28ce49.png)

This image is longer than I thought, so I'm not posting what it looks like in a light colorscheme, but it should look fine as long as vim highlight-groups are properly defined.

Commit:
Added vim syntax files for syntax highlighting

syntax/carbon.vim: highlighting for core Carbon syntax
ftdetect/carbon.vim: filetype detection for .carbon file extension

Commit:
Added README for utils/vim directory

MichaelBonnet commented:
I'm not good enough with things like regex to do any useful review, but stuff like this is great for helping future adoption. Good job.

jonmeow commented on utils/vim/syntax/carbon.vim:
```diff
@@ -0,0 +1,117 @@
+syn keyword carbonBooleanType bool
+syn keyword carbonIntType i8 i16 i32 i64 i128
+syn keyword carbonUnsignedIntType u8 u16 u32 u64 u128
+syn keyword carbonFloatType f16 f32 f64
```
Just noting, these are actually going to be i\d+, u\d+, f\d+ (even though these are definitely the most common forms)

x2w-soda replied:
Thanks for the review, updated.

jonmeow commented on utils/vim/syntax/carbon.vim:
```diff
@@ -0,0 +1,117 @@
+syn keyword carbonFloatType f16 f32 f64
+syn keyword carbonStringType String
+syn keyword carbonBoolean true false
+syn match carbonInt '\<\d\+\>'
```
It might be worth handling `5_000` here, etc. Here's a regex reference you could use:
https://github.com/carbon-language/carbon-lang/blob/trunk/utils/highlightjs/highlightjs_carbon_lang.js#L139

x2w-soda replied:
The reference helped a lot, thanks. Added support for underscores.

jonmeow reviewed (commented):
Thanks! Generally, this looks good, just a couple comments on regexes. Note the highlightjs regexes were developed by chandlerc, so they may offer a fairly reliable source of data for you.

Commit:
update highlighting policy for primitive types

generalize integer and float types.
support underscores in literals and hex literal exponents.

Commit:
use vimscript comments for copyright preamble

pre-commit workflow suggested using # as vimscript comment for the
copyright preamble, but vimscript comments start with double quotes.

jonmeow commented on utils/vim/syntax/carbon.vim:
```diff
@@ -0,0 +1,112 @@
+" Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+" Exceptions. See /LICENSE for license information.
+" SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Sorry, in /.pre-commit-config.yaml, under the copyright check, can you add:

```
          - --custom_format
          - '\.vim$'
          - ''
          - '" '
          - ''
```

That should fix issues with these.

jonmeow commented on utils/vim/README.md:
```diff
@@ -0,0 +1,29 @@
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+# Carbon Syntax Highlighting for Vim & Neovim
```
```suggestion
# Carbon Syntax Highlighting for Vim & Neovim

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

```

Minor, we're just generally trying to put the title at the top of the file.

jonmeow commented on utils/vim/README.md:
```diff
@@ -0,0 +1,29 @@
+
+# Carbon Syntax Highlighting for Vim & Neovim
+
+For Carbon devs using Vim or Neovim, this plugin provides syntax highlighting for .carbon files found throughout `explorer/testdata`
```
Note running this through prettier or pre-commit will do word-wrap for you.

https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#pre-commit
https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#prettier

jonmeow commented on utils/vim/README.md:
```diff
@@ -0,0 +1,29 @@
+
+For Carbon devs using Vim or Neovim, this plugin provides syntax highlighting for .carbon files found throughout `explorer/testdata`
+
+# Manual Installation
+
+## Vim Users
```
```suggestion
## Manual Installation

### Vim Users
```

Noticed these headers are slightly off for md, should only have one top-level

jonmeow commented on utils/vim/README.md:
````diff
@@ -0,0 +1,29 @@
+mkdir -p ~/.vim/ftdetect && cp ftdetect/carbon.vim ~/.vim/ftdetect/
+```
+
+## Neovim Users
````
```suggestion
### Neovim Users
```

jonmeow reviewed (commented):
Looking at this mainly for the pre-comit issues:
https://github.com/carbon-language/carbon-lang/runs/7593471338?check_suite_focus=true

(note, in addition to running commands, you could also just copy-paste the wrap edits suggested there)

x2w-soda committed:
Move README title to top

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

x2w-soda committed:
adjust README header level

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

Commit:
support .vim file copyright check in pre-commit

Commit:
use Prettier for word-wrapping

x2w-soda commented:
Added copyright check support for .vim files in pre-commit workflow. Formatted README as requested, thank you for your patience @jonmeow .

jonmeow committed:
Merge branch 'trunk' into vim-highlight

jonmeow reviewed (approved):
Thanks, this looks good! I'm happy to use this as a starting point for people.

I did a quick conflict resolve on the pre-commit-config, just an overlapping change that'd adding plist.

mattn commented:
Thanks for the Vim plugin But could you please separate files from carbon-language?

x2w-soda commented:
Hi, can you elaborate on what you mean by "separating files from carbon-language"? I'm not quite sure what you mean.

mattn commented:
I just hope that this vim syntax files will be moved to 

https://github.com/carbon-language/vim-carbon-lang

x2w-soda commented:
Oh, I only placed the syntax files in `utils/vim` because other syntax highlighting utilities are also there. You could open an issue to migrate the syntax files to that new repository if you like.

mattn commented:
Now I confirmed vim-carbon-lang is ready to use. Thank you.
